### PR TITLE
DAOS-7130 metrics: Simplify telemetry API

### PR DESCRIPTION
**Pull request by kjacque**
This patch simplifies the API to focus on the common use
cases for the producer and consumer, where they first add
or get the metric before setting or getting values,
respectively.

- Remove path options for metric getters and setters. The
  caller is required to have a valid pointer to the node
  in the metric tree.
- Remove return codes from the setters. There is nothing
  the caller can really do if it fails.
- Consolidate unit tests into a single test file, for
  ease of future maintenance.
- Fix a bug where buckets weren't populated with the
  pointer to the bucket metric node.
- Return a more descriptive code and log when units or
  description strings are too long in d_tm_add_metric.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

File: src/container/SConscript [hunk 1]
```diff
@@ -18,7 +18,8 @@ def scons():
                                  ['srv.c', 'srv_container.c', 'srv_epoch.c',
                                   'srv_target.c', 'srv_layout.c', 'oid_iv.c',
                                   'container_iv.c', 'srv_csum_recalc.c',
-                                  'srv_cli.c', 'srv_oi_table.c', common],
+                                  'srv_cli.c', 'srv_oi_table.c',
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/container/SConscript:21]: Fixed

File: src/gurt/telemetry.c [hunk 2]
```diff
@@ -1038,847 +1038,249 @@ d_tm_compute_stats(struct d_tm_node_t *node, uint64_t value)
...
- * The duration is specified either by an initialized pointer or by a fully
- * qualified item name.  If an initialized pointer is provided, the metric is
- * accessed directly.  Otherwise, a lookup is performed on the path name
- * provided in order to find the specified item.  Because this function must be
- * paired with d_tm_mark_duration_start(), the metric is not created if it
- * does not already exist.
- *
  * \param[in]		metric	Pointer to the metric
- * \param[in]		err	If non-zero, aborts the interval calculation
- *				so that this interval is not added to the stats.
```
Comment (jbrosenz) [inline src/gurt/telemetry.c:1494]: Was the removal of err done on purpose?  That was added at Johann's request not all that long ago.

**Review by jbrosenz — DISMISSED** [review]
The mechanics look fine.  There are no issues there.  I like the simplification that results from no longer specifying a fmt string to implicitly create metrics upon their first use.  Though, I thought that was a desirable feature that the team wanted early on, which is what guided the API to look the way it did.  If the usage data suggests that explicit initialization is the way to go, then this change certainly is good.

File: src/gurt/telemetry.c (hunk 2 above)
Comment (kjacque) [reply src/gurt/telemetry.c:1494]: Yeah, I didn't see it used anywhere outside of tests, and it seemed like an odd pattern. If someone intends to use it, I'll happily add it back in.

File: src/engine/init.c [hunk 3]
```diff
@@ -460,9 +460,8 @@ static void
 dss_crt_event_cb(d_rank_t rank, enum crt_event_source src,
 		 enum crt_event_type type, void *arg)
 {
-	static struct d_tm_node_t	*dead_ranks;
-	static struct d_tm_node_t	*last_ts;
-	int				 rc = 0;
+	int			rc = 0;
+	struct engine_metrics	*metrics;
```
Comment (tanabarr) [inline src/engine/init.c:464]: nit: variable alignment

File: src/engine/init.c [hunk 4]
```diff
@@ -502,10 +504,11 @@ server_id_cb(uint32_t *tid, uint64_t *uid)
...
 {
-	uint64_t	bound;
-	int64_t		diff;
-	unsigned int	ctx_nr;
-	int		rc;
+	uint64_t		bound;
+	int64_t			diff;
+	unsigned int		ctx_nr;
+	int			rc;
+	struct engine_metrics	*metrics;
```
Comment (tanabarr) [inline src/engine/init.c:511]: nit: variable alignment

**Review by tanabarr — APPROVED** [review]

File: src/engine/init.c [hunk 5]
```diff
@@ -521,13 +524,20 @@ server_init(int argc, char *argv[])
...
+
 	/** Report timestamp when engine was started */
-	(void)d_tm_record_timestamp(NULL, "started_at");
+	d_tm_record_timestamp(metrics->started_time);
 
 	rc = drpc_init();
 	if (rc != 0) {
 		D_ERROR("Failed to initialize dRPC: "DF_RC"\n", DP_RC(rc));
-		goto exit_telemetry_init;
+		goto exit_metrics_init;
```
Comment (johannlombardi) [inline src/engine/init.c:540]: As for others, this should not be fatal.

File: src/engine/srv_metrics.c [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+			     "Timestamp of last received event", NULL,
+			     "events/last_event_ts");
+	if (rc != 0) {
+		D_ERROR("unable to add metric for last event timestamp: "
+			DF_RC "\n", DP_RC(rc));
+		return rc;
+	}
+
+	return 0;
+}
```
Comment (johannlombardi) [inline src/engine/srv_metrics.c:71]: I would like to eventually generalize the macros I added to bio. Anyway, we can again fix that in a subsequent patch.

**Review by johannlombardi — CHANGES_REQUESTED** [review]
Thanks. Must merge master now that i have landed the bio patch.

File: src/container/srv_metrics.c [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * Fetch the container module's global metrics.
+ *
+ * \return container metrics structure
+ */
+struct cont_metrics *
+ds_cont_get_metrics(void)
+{
+	return &metrics;
+}
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/container/srv_metrics.c:79]: Changing them to externs.

File: src/engine/init.c (hunk 5 above)
Comment (kjacque) [reply src/engine/init.c:540]: Fixed.

File: src/engine/init.c (hunk 3 above)
Comment (kjacque) [reply src/engine/init.c:464]: Fixed both of these variable alignments

File: src/include/gurt/telemetry_producer.h [hunk 8]
```diff
@@ -6,24 +6,18 @@
...
-			 const char *fmt, ...);
-int d_tm_decrement_gauge(struct d_tm_node_t **metric, uint64_t value,
-			 const char *fmt, ...);
+void d_tm_record_timestamp(struct d_tm_node_t *metric);
+void d_tm_take_timer_snapshot(struct d_tm_node_t *metric, int clk_id);
+void d_tm_mark_duration_start(struct d_tm_node_t *metric, int clk_id);
+void d_tm_mark_duration_end(struct d_tm_node_t *metric);
+void d_tm_set_gauge(struct d_tm_node_t *metric, uint64_t value);
+void d_tm_increment_gauge(struct d_tm_node_t *metric, uint64_t value);
+void d_tm_decrement_gauge(struct d_tm_node_t *metric, uint64_t value);
```
Comment (johannlombardi) [inline src/include/gurt/telemetry_producer.h:20]: I would have used inc_gauge and dec_gauge to be consistent with counter.

Comment (kjacque) [reply src/include/gurt/telemetry_producer.h:20]: Good point. I'll change them.

**Review by tanabarr — APPROVED** [review]

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**